Source/JavaScriptCore/heap/MarkedBlock.cpp

```diff
@@ -502,6 +512,119 @@ void MarkedBlock::Handle::sweep(FreeList* freeList)
+    LOG_INVALID_HANDLE_DETAILS("MarkedBlock = %p ; heapCell = %p ; type = %d\n", this, heapCell, cellType);
+
+    SAVE_TO_REG(savedMarkedBlock, "x19", this);
+    SAVE_TO_REG(savedHeapCell, "x20", heapCell);
+    SAVE_TO_REG(savedType, "x21", cellType);
+
+    static_assert(!offsetOfHeader);
+    static_assert(!OBJECT_OFFSETOF(Header, m_handle));
+    size_t contiguousZeroCountAfterHandle = 0;
+    {
+        char* mem = WTF::bitwise_cast<char*>(&header());
+        for (; contiguousZeroCountAfterHandle < MarkedBlock::blockSize; ++contiguousZeroCountAfterHandle) {
+            if (*mem)
+                break;
+            ++mem;
+        }
+    }
+    LOG_INVALID_HANDLE_DETAILS("found %zd 0s at beginning of block\n", contiguousZeroCountAfterHandle);
+    SAVE_TO_REG(savedCount, "x22", contiguousZeroCountAfterHandle);
+
+    bool isValidBlockVM = false;
+    bool foundBlockInThisVM = false;
+    bool isBlockInVM = false;
+    bool isBlockHandleInVM = false;
+
+    VM* blockVM = header().m_vm;
+    VM* actualVM = header().m_vm;
+    SAVE_TO_REG(savedBlockVM, "x23", blockVM);
+    SAVE_TO_REG(savedActualVM, "x24", actualVM);
+    SAVE_TO_REG(savedBitfield, "x25", 0L);
+
+    {
+        VMInspector::forEachVM([&](VM& vm) {
+            if (blockVM == &vm) {
+                isValidBlockVM = true;
+                savedBitfield = 8;
+                LOG_INVALID_HANDLE_DETAILS("block VM %p is valid\n", &vm);
+                return IterationStatus::Done;
+            }
+            return IterationStatus::Continue;
+        });
+    }
+
+    if (isValidBlockVM) {
+        MarkedSpace& objectSpace = blockVM->heap.objectSpace();
+        isBlockInVM = objectSpace.blocks().set().contains(this);
+        isBlockHandleInVM = !!objectSpace.findMarkedBlockHandleDebug(this);
+        foundBlockInThisVM = isBlockInVM || isBlockHandleInVM;
+        LOG_INVALID_HANDLE_DETAILS("block in our VM = %d, block handle in our VM = %d\n", isBlockInVM, isBlockHandleInVM);
+
+        savedBitfield = (isValidBlockVM ? 8 : 0) | (isBlockInVM ? 4 : 0) | (isBlockHandleInVM ? 2 : 0) | (foundBlockInThisVM ? 1 : 0);
+    }
+    if (!isBlockInVM && !isBlockHandleInVM) {
+        // worst case path
+        VMInspector::forEachVM([&](VM& vm) {
+            MarkedSpace& objectSpace = vm.heap.objectSpace();
+            isBlockInVM = objectSpace.blocks().set().contains(this);
+            isBlockHandleInVM = !!objectSpace.findMarkedBlockHandleDebug(this);
+            // Either of them is true indicates that the block belongs or used to belong to the VM.
+            if (isBlockInVM || isBlockHandleInVM) {
+                actualVM = &vm;
+                LOG_INVALID_HANDLE_DETAILS("block in another VM: %d, block in another VM: %d; other VM is %p\n", isBlockInVM, isBlockHandleInVM, &vm);
+
+                savedActualVM = actualVM;
+                savedBitfield = (isValidBlockVM ? 8 : 0) | (isBlockInVM ? 4 : 0) | (isBlockHandleInVM ? 2 : 0) | (foundBlockInThisVM ? 1 : 0);
```

Comment:
To be consistent, can we also SAVE_TO_REG savedActualVM and savedBitfield here?  You can remove actualVM.  Don't think we need it.